src/librustc/ty/sty.rs

```diff
@@ -2151,64 +2151,60 @@ impl<'tcx> Const<'tcx> {
     }
 
     #[inline]
-    pub fn to_bits(
-        &self,
-        tcx: TyCtxt<'_, '_, 'tcx>,
-        ty: ParamEnvAnd<'tcx, Ty<'tcx>>,
-    ) -> Option<u128> {
-        if self.ty != ty.value {
-            return None;
-        }
-        let ty = tcx.lift_to_global(&ty).unwrap();
-        let size = tcx.layout_of(ty).ok()?.size;
-        self.val.try_to_bits(size)
-    }
-
-    #[inline]
-    pub fn to_ptr(&self) -> Option<Pointer> {
-        self.val.try_to_ptr()
-    }
-
-    #[inline]
-    pub fn assert_bits(
+    pub fn try_eval_bits(
         &self,
         tcx: TyCtxt<'_, '_, '_>,
-        ty: ParamEnvAnd<'tcx, Ty<'tcx>>,
+        ty: Ty<'_>,
     ) -> Option<u128> {
-        assert_eq!(self.ty, ty.value);
+        assert_eq!(self.ty, ty);
+        let tcx = tcx.global_tcx();
         let ty = tcx.lift_to_global(&ty).unwrap();
-        let size = tcx.layout_of(ty).ok()?.size;
-        self.val.try_to_bits(size)
+        // FIXME(eddyb, oli-obk) get the right param_env.
+        let size = tcx.layout_of(ParamEnv::empty().and(ty)).ok()?.size;
+        match self.val {
+            ConstValue::Unevaluated(did, substs) => {
+                let substs = tcx.lift_to_global(&substs).unwrap();
+                let instance = ty::Instance::resolve(tcx, ParamEnv::empty(), did, substs)?;
+                let gid = GlobalId {
+                    instance,
+                    promoted: None,
+                };
+                let evaluated = tcx.const_eval(ParamEnv::empty().and(gid)).ok()?;
+                evaluated.val.try_to_bits(size)
+            },
+            // FIXME(const_generics): try to evaluate generic consts with a given param env?
+            // E.g. when you have an associated constant whose value depends on a generic const
```

Comment:
Not sure what this refers to.

Reply:
Well, there are other "unevaluated" const values (Infer and Param) that could be attempted to be evaluated here.

Reply:
`Infer`, maybe, but doubtful (only makes sense within an inference context, but, really, shouldn't happen w/o lazy normalization support in `ty::relate`).

`Param` is not "unevaluated", it means the `ty::Const` is still generic - there's nothing to replace it with.